File: homework-solutions/Group 14/twosat.py
import sys

def main():
    data = sys.stdin.read().strip().split()
    it = iter(data)
    n = int(next(it))
    m = int(next(it))

    def NOT(x):
        return x + n if x <= n else x - n

    size = 2 * n + 5
    G = [[] for _ in range(size)]
    id_scc = [0] * size
    num = [0] * size
    low = [0] * size
    stack = [0] * size
    
    timeDFS = 0
    scc = 0

    def dfs(u):
        nonlocal timeDFS, scc
        timeDFS += 1
        num[u] = low[u] = timeDFS
        stack[0] += 1
        stack[stack[0]] = u

        for v in G[u]:
            if id_scc[v] != 0:
                continue
            if num[v] == 0:
                dfs(v)
                low[u] = min(low[u], low[v])
            else:
                low[u] = min(low[u], num[v])

        if num[u] == low[u]:
            scc += 1
            while True:
                v = stack[stack[0]]
                stack[0] -= 1
                id_scc[v] = scc
                if v == u:
                    break

    def add_clause_and(u, v):
        G[u].append(v)
        G[v].append(u)

    for _ in range(m):
        t = int(next(it))
        u = int(next(it))
        v = int(next(it))
        if t == 1:
            add_clause_and(u, v)
            add_clause_and(NOT(u), NOT(v))
        else:
            add_clause_and(NOT(u), v)
            add_clause_and(u, NOT(v))

    for i in range(1, 2 * n + 1):
        if id_scc[i] == 0:
            dfs(i)

    valid = True
    for i in range(1, n + 1):
        if id_scc[i] == id_scc[NOT(i)]:
            valid = False

    print("NO" if valid else "YES")


if __name__ == "__main__":
    main()

File: homework-solutions/Group 14/twosat.cpp
#include <bits/stdc++.h>

using namespace std;

const int maxN = 500500;

int n, m;

// Lưu đồ thị G
vector<int> G[maxN << 1];

// Lấy giá trị phủ định của x
int NOT(int x) {
    return x + (x <= n ? n : -n); // -x
}

// Thêm điều kiện u OR v
void add_clause_or(int u, int v) {
    G[NOT(u)].push_back(v); // -u -> v
    G[NOT(v)].push_back(u); // -v -> u
}

// Thêm điều kiện u AND v
void add_clause_and(int u, int v) {
    G[u].push_back(v); // u -> v
    G[v].push_back(u); // v -> u
}

// Tìm thành phần liên thông mạnh
int id[maxN << 1];
int num[maxN << 1], low[maxN << 1];
int timeDFS = 0, scc = 0;
int st[maxN << 1];

void dfs(int u) {
    num[u] = low[u] = ++timeDFS;
    st[++st[0]] = u;
    for(const int& v : G[u]) {
        if(id[v] != 0) continue;
        if(num[v] == 0) {
            dfs(v);
            low[u] = min(low[u], low[v]);
        } else low[u] = min(low[u], num[v]);
    }

    if(num[u] == low[u]) {
        for(++scc; true; ) {
            int v = st[st[0]--];
            id[v] = scc;
            if(v == u) break;
        }
    }
}

int main() {
    cin.tie(0)->sync_with_stdio(0);

    cin >> n >> m;
    for(int i = 1; i <= m; ++i) {
        int t; cin >> t;
        int u, v; cin >> u >> v;
        if(t == 1) {
            // u & v | !u & !v
            add_clause_and(u, v);
            add_clause_and(NOT(u), NOT(v));
        } else {
            // u & !v | u & !v
            add_clause_and(NOT(u), v);
            add_clause_and(u, NOT(v));
        }
    }

    // Thuật toán Tarjan
    for(int i = 1; i <= 2 * n; ++i) {
        if(!id[i]) dfs(i);
    }

    bool answer = 1;
    for(int i = 1; i <= n; ++i) {
        // Kiểm tra điều kiện tồn tại phương án
        if(id[i] == id[NOT(i)]) answer = 0;
    }
    cout << (answer ? "NO\n" : "YES\n");
    return 0;
}
